Added nullable treatment to Animal

jcmatese opened the pull request:
<!-- markdownlint-disable-next-line first-line-heading -->
## Summary Change Description

added nullable foreign key to Protocol from Animal
added Protocol.category, as a classification/filter
added a test to ensure MSRun protocols get the correct category



## Affected Issue Numbers

- Resolves #36 , **_at least in a progressive, step-wise fashion_**

Also, just re-contextualizing the decision-making process for making this _nullable_:
1) prior example files did not have Animal Treatment
2) making it _required_ would likely break prior tests, and or require the lab to reformat the example file already supplied to us
3) the [most recent user specification](https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/36#issuecomment-824089017) was "animals might have _**zero**_ to any number of treatments" and that is if a treatment existed we would accept it as "long-form or "free" text" 

 
## Code Review Notes

No significant changes, as Treatment is nullable.
The **_default protocol.category is MSRun focused_**, though, so any future protocol creation code will want to explicitly pass that coded value.

## Checklist

- [ ] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
- [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)

jcmatese commented:
If this is approved, I can move on to the attribute(s)-application (add the Treatment-containing datafile, and sample loading modifications, unless lance has already done that on the `time_collected` front )

jcmatese commented:
lol, I think I misspelled treatment (!). Let's try again.

lparsons reviewed (changes requested):
I understand that this is an attempt at supporting the treatment variable provided by the lab in a lightweight fashion. However, I think there are a number of things missing from the approach that could cause problems. More defining the proposed DB model in the issue would also be helpful.

1. Additional tests should be implemented to ensure correct behavior, including association a protocol with an Animal and deleting a protocol while an animal is associated with it (should fail imo). 
2. Since we are overloading the Protocol table, there should be restrictions/validations put in place to assure that MSRuns are always associated with MSRun protocol types and Animals are associated with Animal_treatment protocols. There should be tests to ensure these validations work.
3. Setting of a default protocol type seems like an odd choice. I would expect that the type be set explicitly upon creation.
4. The CASCADE on delete does not seem like a safe choice since getting rid of a Protocol would remove all associated Animals. I would prefer if the model threw an exception in this case to prevent that from happening.

jcmatese commented on DataRepo/models.py:
```diff
@@ -67,6 +67,33 @@ def atom_count_in_formula(formula, atom):
+    category = models.CharField(
+        max_length=256,
+        choices=CATEGORY_CHOICES,
+        default=MSRUN_PROTOCOL,
```
I believe the `default` is suggested/used by the Django migration, since it is an added, non-nullable column.  If I remove it, then developer would _likely_ have to provide a value when migrating, and I would have to overhaul the existing code and tests to explicitly set the value.  I don't mind , just stating the alternative ramifications.  Let me know...

lparsons replied:
Thanks, that makes sense. Since it is a non-nullable field, it would have to be set (either during migrations or by recreating the database/migrations). In my opinion that isn't worth what I see as compromising the model design at this point.

jcmatese replied:
I think I will leave the default in this PR, as a convenience to get it set and migrated for developers, and remove it in a subsequent PR based on Issue #94

jcmatese commented on DataRepo/models.py:
```diff
@@ -145,6 +172,15 @@ class Animal(models.Model, TracerLabeledClass):
     studies = models.ManyToManyField(Study, related_name="animals")
+    treatment = models.ForeignKey(
+        Protocol,
+        on_delete=models.CASCADE,
```
I can do that.

jcmatese commented on DataRepo/models.py:
```diff
@@ -145,6 +172,15 @@ class Animal(models.Model, TracerLabeledClass):
+        blank=True,
+        related_name="animals",
+        help_text="Lab controlled label of the actions taken on an animal "
+        "that are not captured by diet, feeding_status, etc.",
```
I can do that.  I just copied/interpreted it from Michael's input (his annotated view of the model), because I think he was thinking it was a field/column that captured  information that was _**not**_ captured by the other columns/attributes of the existing spreadsheet/model.  In practice, though, perhaps all those might be moved anyway, so editing the help test might _make_ it more future-proof (albeit present-ambiguous).  I think I tried for find a medium between definitions like "a flexible identifier for a treatment to an animal (ie drug injection)" from [here](https://docs.google.com/spreadsheets/d/1UI2Fs1hW6RiDcM7JhJYfG6yhecSQ8u7A9vaxSrJGeW0/edit#gid=0) and "an action taken on an animal that is not captured in other variables" from [here](https://github.com/Princeton-LSI-ResearchComputing/tracebase/issues/36#issuecomment-824089017) , in which I transposed "other variables" to "_other defined columns in the file they are providing_" (i.e. diet, feeding_status, etc. ).  Anyway, that was my decision process.

fkang-pu replied:
Non-Blocking:

The proposal described here (#93) is my preferred solution for dealing with animal treatments, taking into account all factors.  I want to add a note that we could use the same approach for "treatment2" (if needed). It just came to mind we could have two foreign keys in Animal referencing Protocol model class. **example:** treatment1 and treatment2, both are foreign keys to Protocol, with different values for **related_nam**e (e.g. treatment1_animals, treatment2_animals).

hepcat72 commented on DataRepo/models.py:
```diff
@@ -67,6 +67,33 @@ def atom_count_in_formula(formula, atom):
+        max_length=256,
+        choices=CATEGORY_CHOICES,
+        default=MSRUN_PROTOCOL,
+        help_text="Classification of the protocol, "
+        "referencing it's intended application",
```
NB. In the current context of this review (having read the issue), this is clear, but thinking down the road, I can see this as being ambiguous (and only elucidated by looking at the values in the database).  I suggest something like:
```suggestion
        help_text="Type of protocol, e.g. an animal treatment or MSRun lab "
        "procedure",
```

jcmatese replied:
I added the specific examples, but did not use "Type" because in fan's proposal, there was a separate `type` attribute, that might still be called into play in the (unknown) future.

jcmatese commented:
I made the easy changes, added a deletion test, and moved some of the comments/requests to #94 , because they would benefit from implementation and model usage. I don't know how easy it will be to restrict/validate protocol.category value based on foreign key (relationship dependency).

lparsons commented:
> I made the easy changes, added a deletion test, and moved some of the comments/requests to #94 , because they would benefit from implementation and model usage. I don't know how easy it will be to restrict/validate protocol.category value based on foreign key (relationship dependency).

Thanks John. I took a stab at adding the validation and some tests for `Animal.treatment` and `MSRun.protocol`. Take a look at PR #96 and merge it into this branch if you're good with it.

jcmatese commented:
> > I made the easy changes, added a deletion test, and moved some of the comments/requests to #94 , because they would benefit from implementation and model usage. I don't know how easy it will be to restrict/validate protocol.category value based on foreign key (relationship dependency).
> 
> Thanks John. I took a stab at adding the validation and some tests for `Animal.treatment` and `MSRun.protocol`. Take a look at PR #96 and merge it into this branch if you're good with it.

Thanks! I merged it and added the explicit msrun `category` in the test and a comment on the validation.  If you think this is good enough for now, I will merge it in and move on,

lparsons reviewed (approved):
Looks good to me, thanks for making that MSRun category explicit, makes it easier to follow.
